File: src/tool/QuickSort.java
package tool;

import java.util.Arrays;

/**
 * quick sort
 */
public class QuickSort {
    public static void main(String[] args) {
        int[] a = new int[]{4, 1, 2, 1, 6};
        sort(a, 0, 4);
        System.out.println(Arrays.toString(a));
    }

    public static void sort(int[] nums, int l, int r) {
        if (nums.length < 2) return;
        if (l < r) {
            int[] tmp = partition(nums, l, r);
            sort(nums, l, tmp[0] - 1);
            sort(nums, tmp[1] + 1, r);
        }
    }

    /**
     * 基础版本快排采用数组最后一个元素作为轴进行划分, 改良版随机选一个轴, 避免6,5,4,3,2,1这样的极端案例使时间复杂度退化成O(n^2).
     */
    private static int[] partition(int[] nums, int l, int r) {
        int i = l - 1, j = r;
        while (l < j) {
            if (nums[l] < nums[r]) {
                swap(nums, ++i, l++);
            } else if (nums[l] == nums[r]) {
                l++;
            } else {
                swap(nums, --j, l);
            }
        }
        swap(nums, r, j);
        return new int[]{i + 1, j};
    }

    private static void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

File: src/sort/third/QuickSort.cpp
#include <iostream>

/**
 * quick sort
 *
 * @since 2020-10-13 Tuesday 13:48 - 14:08
 */
class QuickSort {
    void swap(int *arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    void helper(int *arr, int left, int right) {
        if (left >= right) {
            return;
        }
        int pivot = left + (right - left) / 2;
        swap(arr, pivot, right);
        int small = left - 1, big = right;
        for (int i = left; i < big;) {
            if (arr[i] > arr[right]) {
                swap(arr, i, --big);
            } else if (arr[i] < arr[right]) {
                i++;
                small++;
            } else {
                i++;
            }
        }
        swap(arr, right, big++);
        helper(arr, left, small);
        helper(arr, big, right);
    }

   public:
    void sort(int *arr, int n) {
        helper(arr, 0, n - 1);
    }
};

int main(int argc, char const *argv[]) {
    int arr[7] = {5, 3, 4, 2, 1, 65, 2};
    QuickSort s;
    s.sort(arr, 7);
    for (int i = 0; i < 7; i++) {
        std::cout << arr[i] << " ";
    }
}
